src/sentry/replays/consumers/buffered/lib.py

```diff
@@ -0,0 +1,143 @@
+"""Buffered RunTime implementation.
+
+The Buffering RunTime eagerly processes messages as they are received and waits until its buffer
+is full before flushing those messages. The goal being to achieve efficiencies from batched I/O.
+"""
+
+from collections.abc import Callable, MutableMapping
+from dataclasses import dataclass
+from functools import partial
+from typing import Generic, TypeVar
+
+from arroyo.types import Partition
+
+from sentry.replays.consumers.buffered.platform import (
+    Cmd,
+    Commit,
+    Flags,
+    Join,
+    Poll,
+    RunTime,
+    Sub,
+    Task,
+)
+
+Item = TypeVar("Item")
+T = TypeVar("T")
+
+
+@dataclass
+class Model(Generic[Item]):
+    buffer: list[Item]
+    can_flush: Callable[["Model[Item]"], bool]
+    do_flush: Callable[["Model[Item]"], None]
+    offsets: MutableMapping[Partition, int]
```

Comment:
I think this is problematic, it complicates the architecture and makes mistakes easier.
Arroyo provides offset management out of the box to hide it almost entirely from the application. "almost" means that the user needs to pick how often to commit and to add a Commit strategy at the end of the consumer.

Other than that messages carry offsets, transformed messages carry the same offsets and batches carry watermarks. These are just managed by the apis. If the arroyo primitives are used, the developer never deals with offsets. 

Here we are resurfacing them and we manage offsets and watermarks explicitly. Which means we would be doing the same thing twice.